### ISSUE TITLE
allow claims to be configured to always be included in id_token

### ISSUE THREAD

**Issue by damienbod** [IdentityServer/IdentityServer4#528]

**Comment by brockallen** [comment]
Can you elaborate? Do you have a IdentityResource configured with role user claims?

**Comment by damienbod** [comment]
Using rc3 or older version the id_token using the Implicit Flow returned roles with the claims per user. Updated to rc4 and the roles are no longer returned  in the id_token.

Flow Implicit Flow reference token

Yes everything was correct before and worked without problems. After the update, roles are missing from the id_token.

Tried all types of configs
```
// Copyright (c) Brock Allen & Dominick Baier. All rights reserved.
// Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.

using IdentityServer4.Models;
using System.Collections.Generic;

namespace QuickstartIdentityServer
{
    using System.Security.Claims;

    public class Config
    {
        public static IEnumerable<IdentityResource> GetIdentityResources()
        {
            return new List<IdentityResource>
            {
                new IdentityResources.OpenId(),
                new IdentityResources.Profile(),
                new IdentityResources.Email(),
               // new IdentityResource("role",new []{ "securedFiles", "securedFiles.admin", "securedFiles.user", "dataEventRecords", "dataEventRecords.admin" , "dataEventRecords.user" } )
            };
        }

        public static IEnumerable<ApiResource> GetApiResources()
        {
            return new List<ApiResource>
            {
                new ApiResource("dataEventRecords")
                {
                    ApiSecrets =
                    {
                        new Secret("dataEventRecordsSecret".Sha256())
                    },
                    Scopes =
                    {
                        new Scope
                        {
                            Name = "dataEventRecords",
                            DisplayName = "Scope for the data event records resource.",
                            UserClaims = { "role", "dataEventRecords" }
                        }
                    }
                },
                new ApiResource("securedFiles")
                {
                    ApiSecrets =
                    {
                        new Secret("securedFilesSecret".Sha256())
                    },
                    Scopes =
                    {
                        new Scope
                        {
                            Name = "securedFiles",
                            DisplayName = "Scope for the secured files resource.",
                            UserClaims = { "role", "securedFiles" }
                        }
                    }
                }
            };
        }

        // clients want to access resources (aka scopes)
        public static IEnumerable<Client> GetClients()
        {
            // client credentials client
            return new List<Client>
            {
                new Client
                {
                    ClientName = "angular2client",
                    ClientId = "angular2client",
                    AccessTokenType = AccessTokenType.Reference,
                    //AccessTokenLifetime = 600, // 10 minutes, default 60 minutes
                    AllowedGrantTypes = GrantTypes.Implicit,
                    AllowAccessTokensViaBrowser = true,
                    RedirectUris = new List<string>
                    {
                        "https://localhost:44311"

                    },
                    PostLogoutRedirectUris = new List<string>
                    {
                        "https://localhost:44311/Unauthorized"
                    },
                    AllowedCorsOrigins = new List<string>
                    {
                        "https://localhost:44311",
                        "http://localhost:44311"
                    },
                    AllowedScopes = new List<string>
                    {
                        "openid",
                        "dataEventRecords",
                        "securedFiles",
                        "role"
                    }
                }
            };
        }
    }
}
```

**Comment by brockallen** [comment]
Is your angular client using the user info endpoint?

**Comment by damienbod** [comment]
Looks like the implementation of IProfileService is not getting called anymore, although need to debug this properly.

**Comment by damienbod** [comment]
user info endpoint, no, I had the claims in the id_token, so did not require the user info endpoint

**Comment by brockallen** [comment]
we removed the `AlwaysIncludeInIdToken` (and the `UserClaim` class), thus the change in behavior

**Comment by damienbod** [comment]
So I need to do an extra round trip to get the user claims when using a reference token?

**Comment by brockallen** [comment]
As per the spec, yes. And it's unrelated to the type of access token -- it has to do with the response_type being requested.

**Comment by brockallen** [comment]
And if you're using oidc-client as your client-side protocol library, this is done for you.

**Comment by damienbod** [comment]
It's still an extra round trip, would it not be possible to add it back, better user experience

**Comment by brockallen** [comment]
we opted for removing it because most of the time you want to keep the size of the id_token small for JS-style clients

**Comment by damienbod** [comment]
But the dev should be able to decide, usually I require only specific claims , not the whole user profile, this should be the app developers choice, if extra data should be included in the id_token

**Comment by damienbod** [comment]
A extra round trip is less preferred

**Comment by brockallen** [comment]
understood. you have to convince @leastprivilege -- he's the grumpy one :)

**Comment by damienbod** [comment]
Ok, thanks for your time. This is not an improvement for me. Would love to see this possible again.

**Comment by brockallen** [comment]
just for background -- we wanted to get rid of the `AlwaysIncludeInIdToken` on the `UserClaim`, since `UserClaim` was used in 3 places and only one of them was related to id_token. we wanted to maybe move that to a `Client` configuration, but then we were faced with `ClaimsToIncludeInIdToken` or `ClaimsToExcludeFromIdToken` on the `Client` and that became a mess. So... it's tricky, and we thought that since we rarely saw people using `AlwaysIncludeInIdToken` since the spec focuses on user info endpoint and so many of the libraries call it for you, that we thought we could just go for simplicity and remove it all together.

**Comment by leastprivilege** [comment]
Do you still have the same opinion after your migration is done?

Is it all about the one additional round trip to the server at login time? Or are there other problems.

The main motivation for this is to keep the id_token small if possible. It is transmitted via the URL which has length restrictions and you need to keep it around (e.g. in a cookie in server-based apps).

The spec says:

> The Claims requested by the profile, email, address, and phone scope values are returned from the UserInfo Endpoint, as described in Section 5.3.2, when a response_type value is used that results in an Access Token being issued. However, when no Access Token is issued (which is the case for the response_type value id_token), the resulting Claims are returned in the ID Token.

**Comment by damienbod** [comment]
Yes, I think every extra round trip which is not required is too much. For me the balance between security and user experience must be correct. If I increase the size of the header, url, then it's my job  to decrease the size of the body, but I think as a developer, I should have the choice.

For example working in a train, each round trip and is a pain. (Requirements should decide, not the library.)

The motivation as you said, I totally agree, it's just round trips are also painful in some user cases. 

Greetings Damien

**Comment by leastprivilege** [comment]
And do you have a strong case for configuring the transport mechanism claim by claim - or would a per client setting (e.g. `AlwaysIncludeIdentityClaimsInIdentityToken`) be good enough?

**Comment by dodegaard** [comment]
Adding my two cents as I wondered about the change as well.  Per client setting would suffice for my implementation I currently am attempting. Thanks.

**Comment by damienbod** [comment]
claim by claim would be ok for me. Usually I required only 1 or 2 claims from the whole user profile here.

**Comment by leastprivilege** [comment]
so why do you need it claim by claim then? Why not as a client setting?

**Comment by VsLoureiro** [comment]
I am also facing this issue migrating from rc3 to rc4. I have also one or two claims that I need to get initially using the hybrid flow. Claim by Claim would suit best, but I guess per client would do as well.

**Comment by brockallen** [comment]
I added a `AlwaysIncludeUserClaimsInIdToken` property on the `Client` configuration to achieve this. The property name might change, but the functionality is now in there.

**Comment by aggieben** [comment]
@leastprivilege 
> so why do you need it claim by claim then? Why not as a client setting?
In my case, I agree with your goal of keeping the id_token minimally small in size, but need to include a single claim, not all of them.  Client setting is fine, but I need something more fine-grained than "all or nothing".

**Comment by LindaLawton** [comment]
It would be nice if we could pick which claims go into the tokens.    I am happy to have userinfo show everything.   But we dont need everything in the token.

**Comment by brockallen** [comment]
> It would be nice if we could pick which claims go into the tokens. I am happy to have userinfo show everything. But we dont need everything in the token.

You can by configuring `UserClaims` on your `IdentityResources`: http://docs.identityserver.io/en/release/reference/identity_resource.html#refidentityresource

**Comment by LindaLawton** [comment]
@brockallen thanks for the tip thats awsom i am going to look into setting that up now.    I ended up creating a really ugly linq solution in the mean time.

```
context.IssuedClaims = currentUserClaims.ToList();
            if (context.Caller != IdentityServerConstants.ProfileDataCallers.UserInfoEndpoint)
                context.IssuedClaims = context.IssuedClaims.Where(a => !UserClaimsService.UserClaimsThatShouldNotBeIncludeInToken().Contains(a.Type)).ToList();
```

**Comment by nenadvicentic** [comment]
Is there anyway to do the following in implicit flow (website authentication) in `IdentityServer4`:

1. Include "preferred_username" and "roles" into identity token
2. Load additional claims from userendpoint: "given_name", "family_name" and "birthdate" later, using access token which was passed together .

I am trying to migrate from `IdentityServer3`, where this was trivial to implement, setting `Required = true` property for "preferred_username" and "roles" claims, to push them in the token.

**Comment by brockallen** [comment]
You could control that behavior in your profile service.

**Comment by nenadvicentic** [comment]
@brockallen Can you elaborate shortly on that, please?

As far as I understand, I would need to parse `ProfileDataRequestContext` `Caller` property and then, based to decide based on `Client` which claims to pass and which once to filter?

**Comment by brockallen** [comment]
Yep, something like that.

**Comment by nenadvicentic** [comment]
Hmm... OK. That will require to maintain parallel configuration and to ensure it's always in sync with IdentityServer's configuration of `IdentityResources`. Feels a bit hacky, but I'll give it a shot.

Thank you for the tip. I would not come up with that solution on my own.

**Comment by nenadvicentic** [comment]
@brockallen  Just from curiosity, what was the reason behind the decision to remove that feature from `IdentityServer4`?

**Comment by brockallen** [comment]
https://github.com/IdentityServer/IdentityServer4/issues/528#issuecomment-264591934

**Comment by nenadvicentic** [comment]
So `Required` was 1st renamed to `AlwaysIncludeInIdToken` and then removed. Did not get that while reading 1st time.

Regarding the scenario where user info endpoint is called automatically, immediately after login, client app still ends up persisting all claims permanently. 

My scenario is to call `id_token` with minimal number of claims (`sub` and `roles`), to have less data in the cookie, and then call user-info-endpoint, only when the rest of the data is needed (rarely), without persisting it.

@brockallen Thanks for quick responses and the clarification!

**Comment by nenadvicentic** [comment]
@brockallen I tried to override `IProfileService` service, but solution is less then optimal.

`GetProfileDataAsync` method gets called only if `AlwaysIncludeUserClaimsInIdToken` is set in client configuration. So effectively,  `AlwaysIncludeUserClaimsInIdToken=true` has to be configured, to achieve filtering of claims. Separate (but dependent) custom configuration is also required, to define what to filter, for which client. It becomes very confusing what is actually going on.

I decided to go in other direction, make second call on every client to `user-info-endpoint` in `OnTokenValidated` handler and each application to filter claims received from the endpoint. Much more work, but easier to understand and follow.

I hope we get this feature back in `IdentityServer vNext`.
